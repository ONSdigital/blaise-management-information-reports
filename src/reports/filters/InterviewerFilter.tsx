import {StyledForm} from "blaise-design-system-react-components";
import React, {ReactElement} from "react";
import dateFormatter from "dayjs";
import Breadcrumbs from "../../components/Breadcrumbs";
import CallHistoryLastUpdatedStatus from "../../components/CallHistoryLastUpdatedStatus";

interface InterviewerFilterPageProps {
    title: string
    interviewer: string | undefined
    setInterviewer: (string: string) => void
    startDate: Date
    setStartDate: (Date: Date) => void
    endDate: Date
    setEndDate: (Date: Date) => void
    surveyTla: string | undefined
    setSurveyTla: (string: string) => void
    submitFunction: () => void;
}

function InterviewerFilter(props: InterviewerFilterPageProps): ReactElement {
    const {
        title,
        interviewer,
        setInterviewer,
        startDate,
        setStartDate,
        endDate,
        setEndDate,
        surveyTla,
        setSurveyTla,
        submitFunction
    } = props;

    async function submitInterviewerFilters(formValues: Record<string, any>, setSubmitting: (isSubmitting: boolean) => void): Promise<void> {
        setInterviewer(formValues["Interviewer ID"]);
        setStartDate(formValues["Start date"]);
        setEndDate(formValues["End date"]);
        setSurveyTla(formValues["Survey TLA"]);
        setSubmitting(true);
        submitFunction();
    }

    const validateInterviewer = (value: string) => {
        let error;
        if (value === "" || value === undefined) {
            error = "Enter a interviewer ID";
        }
        return error;
    };

    const validateDate = (value: string) => {
        let error;
        if (value === "" || value === undefined) {
            error = "Enter a date";
            return error;
        }

        if (value.match(/^[0-9]{4}-[0-9]{2}-[0-9]{2}$/) === null) {
            error = "Enter a valid date";
        }
        return error;
    };

    const fields = [
        {
            name: "Survey TLA",
            description: "Select survey",
            type: "radio",
            initial_value: surveyTla,
            radioOptions: [
                {id: "all", value: "", label: "Show all surveys"},
                {id: "lms", value: "lms", label: "LMS", description: "Labour Market Survey"},
                {id: "opn", value: "opn", label: "OPN", description: "Opinions and Lifestyle Survey"}
            ]
        },
        {
            name: "Interviewer ID",
            type: "text",
            initial_value: interviewer,
            validate: validateInterviewer
        },
        {
            name: "Start date",
            type: "date",
            initial_value: dateFormatter(startDate).format("YYYY-MM-DD"),
            validate: validateDate
        },
        {
            name: "End date",
            type: "date",
            initial_value: dateFormatter(endDate).format("YYYY-MM-DD"),
            validate: validateDate
        }
    ];

    return (
        <>
            <div>
                <Breadcrumbs
                    BreadcrumbList={[{link: "/", title: "Reports"}]}/>
                <main id="main-content" className="page__main u-mt-s">
                    <h1 className="u-mb-m">Run interviewer {title} report</h1>
                    <CallHistoryLastUpdatedStatus/>
                    <StyledForm fields={fields} onSubmitFunction={submitInterviewerFilters} submitLabel={"Next"}/>
                </main>
            </div>
        </>
    );
}

export default InterviewerFilter;


