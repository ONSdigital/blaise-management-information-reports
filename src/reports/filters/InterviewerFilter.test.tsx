/**
 * @jest-environment jsdom
 */

import "@testing-library/jest-dom";
import { createMemoryHistory } from "history";
import { render, RenderResult, waitFor } from "@testing-library/react";
import { Router } from "react-router";
import { screen } from "@testing-library/dom";
import React from "react";
import InterviewerFilter, { InterviewerFilterQuery } from "./InterviewerFilter";
import userEvent from "@testing-library/user-event";

describe("the interviewer details page renders correctly", () => {

    let submitFunction: () => void;

    let view: RenderResult;

    beforeEach(() => {
        submitFunction = jest.fn();

        const history = createMemoryHistory();

        const initialQuery: InterviewerFilterQuery = {
            interviewer: "James",
            startDate: new Date("2021-01-01"),
            endDate: new Date("2021-01-05"),
            surveyTla: "LMS",
        };

        view = render(
            <Router history={history}>
                <InterviewerFilter
                    title=""
                    query={initialQuery}
                    onSubmit={submitFunction} />
            </Router>
        );
    });

    it("matches snapshot", async () => {
        jest.useRealTimers();

        expect(await view).toMatchSnapshot();
    });

    it("renders correctly", async () => {
        expect(screen.getByText(/Run interviewer/i)).toBeVisible();
        expect(screen.getByText(/Data in this report was last updated:/i)).toBeVisible();

        expect(screen.getByText(/Select survey/i)).toBeVisible();
        expect(screen.getByText(/Show all surveys/i)).toBeVisible();
        expect(screen.getByText(/LMS/i)).toBeVisible();
        expect(screen.getByText(/OPN/i)).toBeVisible();

        expect(screen.getByText(/Interviewer ID/i)).toBeVisible();

        expect(screen.getByText(/Start date/i)).toBeVisible();
        expect(screen.getByText(/End date/i)).toBeVisible();
    });

    it("updates the values", async () => {
        const interviewer = screen.getByRole("textbox", { name: "Interviewer ID" });
        userEvent.clear(interviewer);
        userEvent.type(interviewer, "rich");

        const startDate = screen.getByLabelText("Start date");
        userEvent.clear(startDate);
        userEvent.type(startDate, "2022-10-10");

        const endDate = screen.getByLabelText("End date");
        userEvent.clear(endDate);
        userEvent.type(endDate, "2022-10-11");

        const survey = screen.getByRole("radio", { name: "LMS Labour Market Survey" });
        userEvent.click(survey);

        userEvent.click(screen.getByRole("button", { name: "Next" }));

        await waitFor(() => {
            expect(submitFunction).toHaveBeenCalledWith({
                interviewer: "rich",
                startDate: new Date("2022-10-10"),
                endDate: new Date("2022-10-11"),
                surveyTla: "lms",
            });
        });
    });
});
