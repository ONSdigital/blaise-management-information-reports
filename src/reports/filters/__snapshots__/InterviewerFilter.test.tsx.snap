// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the interviewer details page renders correctly matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <nav
          aria-label="Breadcrumb"
          class="ons-breadcrumb"
        >
          <ol
            class="ons-breadcrumb__items ons-u-fs-s"
          >
            <li
              class="ons-breadcrumb__item"
              id="breadcrumb-0"
            >
              <a
                class="ons-breadcrumb__link"
                href="/"
              >
                Reports
              </a>
              <svg
                class="ons-svg-icon"
                focusable="false"
                viewBox="0 0 8 13"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
                  transform="translate(-5.02 -1.59)"
                />
              </svg>
            </li>
          </ol>
        </nav>
        <main
          class="ons-page__main ons-u-mt-s"
          id="main-content"
        >
          <h1
            class="ons-u-mb-m"
          >
            Run interviewer 
             report
          </h1>
          <p
            aria-live="polite"
            class="ons-u-fs-s u-mt-s"
          >
            Data in this report was last updated: 
            <b>
              Loading
            </b>
          </p>
          <p
            class="ons-u-fs-s"
          >
            Data in this report only goes back to the last 12 months.
          </p>
          <form
            action="#"
          >
            <fieldset
              class="ons-fieldset"
            >
              <legend
                class="ons-fieldset__legend"
              >
                Select survey
              </legend>
              <div
                class="ons-radios__items"
                id="Survey TLA"
              >
                <p
                  class="ons-radios__item"
                >
                  <span
                    class="ons-radio"
                  >
                    <input
                      class="ons-radio__input js-radio"
                      id="all"
                      initial_value="LMS"
                      name="Survey TLA"
                      type="radio"
                      value=""
                    />
                    <label
                      class="ons-radio__label "
                      for="all"
                      id="all-label"
                    >
                      Show all surveys
                    </label>
                  </span>
                </p>
                <br />
                <p
                  class="ons-radios__item"
                >
                  <span
                    class="ons-radio"
                  >
                    <input
                      class="ons-radio__input js-radio"
                      id="lms"
                      initial_value="LMS"
                      name="Survey TLA"
                      type="radio"
                      value="lms"
                    />
                    <label
                      class="ons-radio__label ons-label--with-description"
                      for="lms"
                      id="lms-label"
                    >
                      LMS
                      <span
                        class="ons-label__description ons-radio__label--with-description"
                        id="white-label-description-hint"
                      >
                        Labour Market Survey
                      </span>
                    </label>
                  </span>
                </p>
                <br />
                <p
                  class="ons-radios__item"
                >
                  <span
                    class="ons-radio"
                  >
                    <input
                      class="ons-radio__input js-radio"
                      id="opn"
                      initial_value="LMS"
                      name="Survey TLA"
                      type="radio"
                      value="opn"
                    />
                    <label
                      class="ons-radio__label ons-label--with-description"
                      for="opn"
                      id="opn-label"
                    >
                      OPN
                      <span
                        class="ons-label__description ons-radio__label--with-description"
                        id="white-label-description-hint"
                      >
                        Opinions and Lifestyle Survey
                      </span>
                    </label>
                  </span>
                </p>
                <br />
              </div>
            </fieldset>
            <div
              class="ons-field"
            >
              <label
                class="ons-label "
                for="Interviewer ID"
              >
                Interviewer ID
              </label>
              <input
                id="Interviewer ID"
                initial_value="James"
                name="Interviewer ID"
                type="text"
                value="James"
              />
            </div>
            <div
              class="ons-field"
            >
              <label
                class="ons-label "
                for="Start date"
              >
                Start date
              </label>
              <input
                id="Start date"
                initial_value="2021-01-01"
                name="Start date"
                type="date"
                value="2021-01-01"
              />
            </div>
            <div
              class="ons-field"
            >
              <label
                class="ons-label "
                for="End date"
              >
                End date
              </label>
              <input
                id="End date"
                initial_value="2021-01-05"
                name="End date"
                type="date"
                value="2021-01-05"
              />
            </div>
            <br />
            <button
              class="ons-btn "
              data-testid="submit-button"
              type="submit"
            >
              <span
                class="ons-btn__inner"
              >
                Next
              </span>
            </button>
          </form>
        </main>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <nav
        aria-label="Breadcrumb"
        class="ons-breadcrumb"
      >
        <ol
          class="ons-breadcrumb__items ons-u-fs-s"
        >
          <li
            class="ons-breadcrumb__item"
            id="breadcrumb-0"
          >
            <a
              class="ons-breadcrumb__link"
              href="/"
            >
              Reports
            </a>
            <svg
              class="ons-svg-icon"
              focusable="false"
              viewBox="0 0 8 13"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
                transform="translate(-5.02 -1.59)"
              />
            </svg>
          </li>
        </ol>
      </nav>
      <main
        class="ons-page__main ons-u-mt-s"
        id="main-content"
      >
        <h1
          class="ons-u-mb-m"
        >
          Run interviewer 
           report
        </h1>
        <p
          aria-live="polite"
          class="ons-u-fs-s u-mt-s"
        >
          Data in this report was last updated: 
          <b>
            Loading
          </b>
        </p>
        <p
          class="ons-u-fs-s"
        >
          Data in this report only goes back to the last 12 months.
        </p>
        <form
          action="#"
        >
          <fieldset
            class="ons-fieldset"
          >
            <legend
              class="ons-fieldset__legend"
            >
              Select survey
            </legend>
            <div
              class="ons-radios__items"
              id="Survey TLA"
            >
              <p
                class="ons-radios__item"
              >
                <span
                  class="ons-radio"
                >
                  <input
                    class="ons-radio__input js-radio"
                    id="all"
                    initial_value="LMS"
                    name="Survey TLA"
                    type="radio"
                    value=""
                  />
                  <label
                    class="ons-radio__label "
                    for="all"
                    id="all-label"
                  >
                    Show all surveys
                  </label>
                </span>
              </p>
              <br />
              <p
                class="ons-radios__item"
              >
                <span
                  class="ons-radio"
                >
                  <input
                    class="ons-radio__input js-radio"
                    id="lms"
                    initial_value="LMS"
                    name="Survey TLA"
                    type="radio"
                    value="lms"
                  />
                  <label
                    class="ons-radio__label ons-label--with-description"
                    for="lms"
                    id="lms-label"
                  >
                    LMS
                    <span
                      class="ons-label__description ons-radio__label--with-description"
                      id="white-label-description-hint"
                    >
                      Labour Market Survey
                    </span>
                  </label>
                </span>
              </p>
              <br />
              <p
                class="ons-radios__item"
              >
                <span
                  class="ons-radio"
                >
                  <input
                    class="ons-radio__input js-radio"
                    id="opn"
                    initial_value="LMS"
                    name="Survey TLA"
                    type="radio"
                    value="opn"
                  />
                  <label
                    class="ons-radio__label ons-label--with-description"
                    for="opn"
                    id="opn-label"
                  >
                    OPN
                    <span
                      class="ons-label__description ons-radio__label--with-description"
                      id="white-label-description-hint"
                    >
                      Opinions and Lifestyle Survey
                    </span>
                  </label>
                </span>
              </p>
              <br />
            </div>
          </fieldset>
          <div
            class="ons-field"
          >
            <label
              class="ons-label "
              for="Interviewer ID"
            >
              Interviewer ID
            </label>
            <input
              id="Interviewer ID"
              initial_value="James"
              name="Interviewer ID"
              type="text"
              value="James"
            />
          </div>
          <div
            class="ons-field"
          >
            <label
              class="ons-label "
              for="Start date"
            >
              Start date
            </label>
            <input
              id="Start date"
              initial_value="2021-01-01"
              name="Start date"
              type="date"
              value="2021-01-01"
            />
          </div>
          <div
            class="ons-field"
          >
            <label
              class="ons-label "
              for="End date"
            >
              End date
            </label>
            <input
              id="End date"
              initial_value="2021-01-05"
              name="End date"
              type="date"
              value="2021-01-05"
            />
          </div>
          <br />
          <button
            class="ons-btn "
            data-testid="submit-button"
            type="submit"
          >
            <span
              class="ons-btn__inner"
            >
              Next
            </span>
          </button>
        </form>
      </main>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
